pyfr/solvers/euler/kernels/entropyfilter.mako

```diff
@@ -0,0 +1,167 @@
+# -*- coding: utf-8 -*-
+<%inherit file='base'/>
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+<%include file='pyfr.solvers.euler.kernels.entropy'/>
+
+<% inf = 1e20 %>
+
+<%pyfr:macro name='get_minima' params='u, dmin, pmin, emin'>
+    fpdtype_t d, p, e;
+    fpdtype_t ui[${nvars}];
+
+    dmin = ${inf}; pmin = ${inf}; emin = ${inf};
+
+    for (int i = 0; i < ${nupts}; i++)
+    {
+        % for j in range(nvars):
+        ui[${j}] = u[i][${j}];
+        % endfor
+
+        ${pyfr.expand('compute_entropy', 'ui', 'd', 'p', 'e')};
+        dmin = fmin(dmin, d); pmin = fmin(pmin, p); emin = fmin(emin, e);
+    }
+
+    // If enforcing constraints on fpts/qpts, compute minima on fpts/qpts
+    % if con_fpts:
+    for (int i = 0; i < ${nfpts}; i++)
+    {
+        for (int j = 0; j < ${nvars}; j++)
+        {
+            ui[j] = ${pyfr.dot('intfpts[i][{k}]*u[{k}][j]', k=nupts)};
+        }
+
+        ${pyfr.expand('compute_entropy', 'ui', 'd', 'p', 'e')};
+        // Enforce only positivity constraints
+        dmin = fmin(dmin, d); pmin = fmin(pmin, p);
+    }
+    % endif
+
+    % if con_qpts:
+    for (int i = 0; i < ${nqpts}; i++)
+    {
+        for (int j = 0; j < ${nvars}; j++)
+        {
+            ui[j] = ${pyfr.dot('intqpts[i][{k}]*u[{k}][j]', k=nupts)};
+        }
+
+        ${pyfr.expand('compute_entropy', 'ui', 'd', 'p', 'e')};
+        // Enforce only positivity constraints
+        dmin = fmin(dmin, d); pmin = fmin(pmin, p);
+    }
+    % endif
+</%pyfr:macro>
+
+<%pyfr:macro name='apply_filter' params='umodes, vdm, uf, f'>
+    // Compute filtered solution
+    for (int uidx = 0; uidx < ${nupts}; uidx++)
+    {
+        for (int vidx = 0; vidx < ${nvars}; vidx++)
+        {
+            // Use exp(-zeta*ubdegs2) = f**ubdegs2
+            uf[uidx][vidx] = ${' + '.join('vdm[uidx][{k}]*umodes[{k}][vidx]*pow(f, {ubd2})'.format(k=k, ubd2=ubdegs2[k])
```

Comment:
Integer pow is notoriously badly optimised by compilers.  Horner the expression yourself, polyfit (https://github.com/PyFR/PyFR/blob/0a51cd53c7bcc614c5ccf453583af0040cf22894/pyfr/backends/base/makoutil.py#L56) shows how this can be done once you know the indices.

Reply:
In the future, max order might get changed to L2 (or at least have that option added) so I think it's better to leave as is.

Reply:
Thing is, if the pow does not get optimised it is going to be super expensive.  What should be ~10 muls will potentially end up as 125 pow calls each doing ~200 FLOPs.  Best case, even ~10 pow calls can be expensive.  This is not counting the additional multiplies: `pow(f, 3)*a + pow(f, 3)*b + pow(f, 3)*c` will, in general *not* be optimised to `pow(3, f)*(a + b + c)`.  So, even if the `pow`'s are handled well, you still pay the cost of the extra multiplies.  Even if the L2 case while one can not Horner, you can still factor since many terms will likely share the same filter factor.  This may be a reasonable common ground (factored expression by filter factor, with the factor computed through https://en.wikipedia.org/wiki/Addition-chain_exponentiation).  Simpler than Horner as you just need to group a bunch of terms together.

Finally, `pow` with integer arguments breaks frequently with CUDA. See https://stackoverflow.com/questions/64797767/pow-is-not-working-properly-inside-a-device-function-in-cuda  Seems to be pot luck if it will work in a given release or not.